neuralpp/symbolic/basic_interpreter.py

```diff
@@ -20,7 +20,7 @@ def eval(self, expression: Expression, context: Expression):
             # there's three cases of `function` that BasicInterpreter can eval():
             # 1. a Python callable, which we'll directly call
             # 2. an uninterpreted function, which is not callable. We raise Error when encounter it.
-            case FunctionApplication(function=Constant(value=python_callable), arguments=args):
+            case FunctionApplication(function=Constant(value=python_callable), arguments=args, type=_):
```

Comment:
I'll revert this change. It is unnecessary, since if `type=_` we can just omit it.

We can pattern match any @property or attribute of a class. The `__match_args__` of a class just serves to facilitate [Matching positional attributes](https://peps.python.org/pep-0636/#matching-positional-attributes). So maybe I should also add `type` to `__match_args__`.

Reply:
I guess no harm leaving it there if it enables matching without specifying attributes.